transport/src/main/java/io/netty/channel/CoalescingBufferQueue.java

```diff
@@ -0,0 +1,181 @@
+     * @param aggregatePromise used to aggregate the promises for the constituent buffers.
+     * @return a {@link ByteBuf} composed of the enqueued buffers.
+     */
+    public ByteBuf remove(int bytes, ChannelPromise aggregatePromise) {
+        if (bytes < 0) {
+            throw new IllegalArgumentException("bytes cannot be less than 0 : " + bytes);
+        }
+        ObjectUtil.checkNotNull(aggregatePromise, "aggregatePromise");
+
+        // Use isEmpty rather than readableBytes==0 as we may have a promise associated with an empty buffer.
+        if (promisesAndBufs.isEmpty()) {
+            return Unpooled.EMPTY_BUFFER;
+        }
+
+        ByteBuf toReturn = null;
+        CompositeByteBuf composite = null;
+        int originalBytes = bytes;
+        for (;;) {
+            Object entry = promisesAndBufs.poll();
+            if (entry == null) {
+                break;
+            } else if (entry instanceof ChannelPromise) {
+                aggregatePromise.addListener(new ChannelPromiseNotifier((ChannelPromise) entry));
+            } else if (bytes > 0) {
+                ByteBuf entryBuffer = (ByteBuf) entry;
+                int entryReadableBytes = entryBuffer.readableBytes();
+                if (toReturn == null) {
+                    if (entryReadableBytes > bytes) {
+                        promisesAndBufs.addFirst(entryBuffer.retain());
+                        toReturn = entryBuffer.readSlice(bytes);
+                        bytes = 0;
+                    } else if (entryReadableBytes == bytes) {
+                        toReturn = entryBuffer;
+                        bytes = 0;
+                    } else {
+                        composite = alloc.compositeBuffer(promisesAndBufs.size() + 1);
+                        composite.addComponent(entryBuffer);
+                        composite.writerIndex(entryReadableBytes);
+                        bytes -= entryReadableBytes;
+                        toReturn = composite;
+                    }
+                } else {
+                    if (entryReadableBytes > bytes) {
+                        ByteBuf slice = entryBuffer.readSlice(bytes);
+                        composite.addComponent(slice);
+                        composite.writerIndex(composite.writerIndex() + slice.readableBytes());
+                        // The remaining buffer is kept as a retained slice of the original.
+                        promisesAndBufs.addFirst(entryBuffer.slice().retain());
+                        bytes = 0;
+                    } else {
+                        composite.addComponent(entryBuffer);
+                        composite.writerIndex(composite.writerIndex() + entryReadableBytes);
+                        bytes -= entryReadableBytes;
+                    }
+                }
+            } else {
+                // Dequeued an entry we did not consume so add it back at the head of the queue.
+                promisesAndBufs.addFirst(entry);
```

Comment:
What if the head of queue is an empty buffer and `bytes=0`? I don't think we should add it back in that case ... or am I missing something?